File: api/scripts/morning-cron.ts
// Morning cron job — re-enrich N oldest accounts + source N new ones.
//
// Usage:
//   npm run -w api morning:dry              # log the plan, don't call Claude
//   npm run -w api morning:run              # full run with default counts (5+5)
//   MORNING_CRON_REENRICH=10 MORNING_CRON_SOURCE=10 npm run -w api morning:run
//
// Counts default to 5+5 to keep daily cost ~$1.50. Override via env vars
// when you want to ramp up. In Step 6 this is wired to Render's cron.

import { db } from '../src/db.js';
import { logger } from '../src/lib/logger.js';
import { enrichAccount, sourceAccounts } from '../src/lib/enrichment/index.js';

interface MorningCronOptions {
  reEnrichCount: number;
  sourceCount: number;
  enrichNew: boolean;
  dryRun: boolean;
}

interface CronStats {
  reEnrichRequested: number;
  reEnrichSucceeded: number;
  reEnrichSkipped: number;
  reEnrichFailed: number;
  sourceCreated: number;
  sourceSkipped: number;
  enrichNewSucceeded: number;
  enrichNewFailed: number;
  elapsedSeconds: number;
}

async function morningCron(opts: MorningCronOptions): Promise<CronStats> {
  const startedAt = Date.now();
  logger.info({ opts }, 'morning cron starting');

  const stats: CronStats = {
    reEnrichRequested: 0,
    reEnrichSucceeded: 0,
    reEnrichSkipped: 0,
    reEnrichFailed: 0,
    sourceCreated: 0,
    sourceSkipped: 0,
    enrichNewSucceeded: 0,
    enrichNewFailed: 0,
    elapsedSeconds: 0,
  };

  // ── Phase 1: re-enrich the N oldest accounts ──────────────────
  // nulls:'first' means never-enriched accounts get priority.
  if (opts.reEnrichCount > 0) {
    const candidates = await db.account.findMany({
      select: { id: true, name: true, lastEnrichedAt: true },
      orderBy: [{ lastEnrichedAt: { sort: 'asc', nulls: 'first' } }],
      take: opts.reEnrichCount,
    });
    logger.info(
      { count: candidates.length, names: candidates.map((c) => c.name) },
      'phase 1: re-enrichment',
    );
    for (const acct of candidates) {
      stats.reEnrichRequested++;
      if (opts.dryRun) {
        logger.info(
          { accountId: acct.id, name: acct.name, lastEnrichedAt: acct.lastEnrichedAt },
          '[dry-run] would re-enrich',
        );
        continue;
      }
      try {
        const result = await enrichAccount(acct.id, { triggeredBy: 'cron' });
        if (result.skipped) {
          stats.reEnrichSkipped++;
          logger.info({ name: acct.name, reason: result.reason }, 'skipped');
        } else {
          stats.reEnrichSucceeded++;
          logger.info(
            { name: acct.name, fieldsUpdated: result.fieldsUpdated, confidence: result.confidence },
            'enriched',
          );
        }
      } catch (err) {
        stats.reEnrichFailed++;
        logger.error(
          { name: acct.name, err: err instanceof Error ? err.message : String(err) },
          're-enrichment failed',
        );
      }
    }
  }

  // ── Phase 2: source N new accounts ────────────────────────────
  let newAccountIds: string[] = [];
  if (opts.sourceCount > 0) {
    logger.info({ count: opts.sourceCount }, 'phase 2: sourcing');
    if (opts.dryRun) {
      logger.info({ count: opts.sourceCount }, '[dry-run] would source new accounts');
    } else {
      try {
        const result = await sourceAccounts({ count: opts.sourceCount, triggeredBy: 'cron' });
        stats.sourceCreated = result.created.length;
        stats.sourceSkipped = result.skipped.length;
        newAccountIds = result.created.map((a) => a.id);
        logger.info(
          { created: result.created.length, skipped: result.skipped.length, names: result.created.map((a) => a.name) },
          'sourcing complete',
        );
      } catch (err) {
        logger.error(
          { err: err instanceof Error ? err.message : String(err) },
          'sourcing failed',
        );
      }
    }
  }

  // ── Phase 3: enrich the just-sourced accounts ─────────────────
  if (opts.enrichNew && newAccountIds.length > 0) {
    logger.info({ count: newAccountIds.length }, 'phase 3: enriching new accounts');
    for (const id of newAccountIds) {
      try {
        await enrichAccount(id, { force: true, triggeredBy: 'cron' });
        stats.enrichNewSucceeded++;
      } catch (err) {
        stats.enrichNewFailed++;
        logger.error(
          { accountId: id, err: err instanceof Error ? err.message : String(err) },
          'new-account enrichment failed',
        );
      }
    }
  }

  stats.elapsedSeconds = Math.round((Date.now() - startedAt) / 1000);
  return stats;
}

// ── CLI entry point ─────────────────────────────────────────────

const args = new Set(process.argv.slice(2));
const isDryRun = args.has('--dry-run');

const reEnrichCount = parseInt(process.env.MORNING_CRON_REENRICH ?? '5', 10);
const sourceCount = parseInt(process.env.MORNING_CRON_SOURCE ?? '5', 10);
const enrichNew = (process.env.MORNING_CRON_ENRICH_NEW ?? 'true').toLowerCase() !== 'false';

morningCron({ reEnrichCount, sourceCount, enrichNew, dryRun: isDryRun })
  .then((stats) => {
    console.log('');
    console.log('━━━━━━━━━━━━━━━━ Morning Cron Summary ━━━━━━━━━━━━━━━━');
    console.table(stats);
    process.exit(0);
  })
  .catch((err) => {
    logger.fatal({ err: err instanceof Error ? err.message : err }, 'morning cron crashed');
    process.exit(1);
  })
  .finally(() => {
    void db.$disconnect();
  });

File: api-py/crm/management/commands/morning_cron.py
"""
Morning cron job — re-enrich N oldest accounts + source N new ones.

Run manually:
    python manage.py morning_cron                # full run with defaults
    python manage.py morning_cron --dry-run      # log the plan, no Claude calls
    REENRICH=10 SOURCE=10 python manage.py morning_cron

In production, schedule via AWS EventBridge → ECS RunTask (or Lambda
that invokes the management command via the Django app entry point).
See api-py/DEPLOY-AWS.md for the recipe.
"""

from __future__ import annotations

import logging
import os
import time
from dataclasses import asdict, dataclass

from django.core.management.base import BaseCommand
from django.db.models import F

from ai.enrichment import enrich_account
from ai.sourcing import source_accounts
from crm.models import Account

logger = logging.getLogger(__name__)


@dataclass
class CronStats:
    re_enrich_requested: int = 0
    re_enrich_succeeded: int = 0
    re_enrich_skipped: int = 0
    re_enrich_failed: int = 0
    source_created: int = 0
    source_skipped: int = 0
    enrich_new_succeeded: int = 0
    enrich_new_failed: int = 0
    elapsed_seconds: int = 0


class Command(BaseCommand):
    help = "Re-enrich oldest N accounts + source N new ICP-matching accounts."

    def add_arguments(self, parser):
        parser.add_argument(
            "--dry-run",
            action="store_true",
            help="Log the plan without calling Claude. Free.",
        )
        parser.add_argument(
            "--reenrich",
            type=int,
            default=int(os.getenv("MORNING_CRON_REENRICH", "5")),
            help="Number of oldest accounts to re-enrich (default 5).",
        )
        parser.add_argument(
            "--source",
            type=int,
            default=int(os.getenv("MORNING_CRON_SOURCE", "5")),
            help="Number of new accounts to source (default 5).",
        )
        parser.add_argument(
            "--no-enrich-new",
            dest="enrich_new",
            action="store_false",
            default=os.getenv("MORNING_CRON_ENRICH_NEW", "true").lower() != "false",
            help="Skip enriching the just-sourced accounts.",
        )

    def handle(self, *args, **opts):
        dry_run = opts["dry_run"]
        re_enrich_count = opts["reenrich"]
        source_count = opts["source"]
        enrich_new = opts["enrich_new"]

        started = time.monotonic()
        stats = CronStats()
        self.stdout.write(self.style.NOTICE(
            f"morning cron starting (dry_run={dry_run}, "
            f"reenrich={re_enrich_count}, source={source_count}, enrich_new={enrich_new})"
        ))

        # ── Phase 1: re-enrich oldest accounts ────────────────────
        if re_enrich_count > 0:
            # nulls first via Django's nulls_first ordering helper.
            candidates = list(
                Account.objects.order_by(F("last_enriched_at").asc(nulls_first=True))[:re_enrich_count]
            )
            self.stdout.write(f"phase 1: re-enrichment ({len(candidates)} candidates)")
            for acct in candidates:
                stats.re_enrich_requested += 1
                if dry_run:
                    self.stdout.write(
                        f"  [dry-run] would re-enrich {acct.name} "
                        f"(last_enriched={acct.last_enriched_at})"
                    )
                    continue
                try:
                    result = enrich_account(acct.id, triggered_by="cron")
                    if result.skipped:
                        stats.re_enrich_skipped += 1
                        self.stdout.write(f"  skipped {acct.name}: {result.reason}")
                    else:
                        stats.re_enrich_succeeded += 1
                        self.stdout.write(self.style.SUCCESS(
                            f"  enriched {acct.name}: fields={result.fields_updated} "
                            f"conf={result.confidence:.2f}"
                        ))
                except Exception as exc:
                    stats.re_enrich_failed += 1
                    self.stderr.write(self.style.ERROR(f"  failed {acct.name}: {exc}"))

        # ── Phase 2: source N new accounts ───────────────────────
        new_account_ids: list[int] = []
        if source_count > 0:
            self.stdout.write(f"phase 2: sourcing ({source_count} requested)")
            if dry_run:
                self.stdout.write(f"  [dry-run] would source {source_count} new accounts")
            else:
                try:
                    src = source_accounts(count=source_count, triggered_by="cron")
                    stats.source_created = len(src.created)
                    stats.source_skipped = len(src.skipped)
                    new_account_ids = [a.id for a in src.created]
                    self.stdout.write(self.style.SUCCESS(
                        f"  created {len(src.created)} ({', '.join(a.name for a in src.created)}) "
                        f"/ skipped {len(src.skipped)}"
                    ))
                except Exception as exc:
                    self.stderr.write(self.style.ERROR(f"  sourcing failed: {exc}"))

        # ── Phase 3: enrich the just-sourced accounts ────────────
        if enrich_new and new_account_ids:
            self.stdout.write(f"phase 3: enriching {len(new_account_ids)} new accounts")
            for acct_id in new_account_ids:
                try:
                    enrich_account(acct_id, force=True, triggered_by="cron")
                    stats.enrich_new_succeeded += 1
                except Exception as exc:
                    stats.enrich_new_failed += 1
                    self.stderr.write(self.style.ERROR(
                        f"  new-account enrichment failed (id={acct_id}): {exc}"
                    ))

        stats.elapsed_seconds = int(time.monotonic() - started)

        # ── Summary ──────────────────────────────────────────────
        self.stdout.write("")
        self.stdout.write(self.style.NOTICE("━━━ Morning Cron Summary ━━━"))
        for key, value in asdict(stats).items():
            self.stdout.write(f"  {key:<24}{value}")
